packages/zoe/test/unitTests/contracts/loan/test-borrow.js

```diff
@@ -345,6 +349,47 @@ test('getDebtNotifier with interest', async t => {
   });
 });
 
+test('aperiodic interest', async t => {
+  const {
+    borrowFacet,
+    maxLoan,
+    periodUpdater,
+    loanKit,
+  } = await setupBorrowFacet(100000, 40000);
+  // @ts-ignore
+  periodUpdater.updateState(0);
+
+  const debtNotifier = await E(borrowFacet).getDebtNotifier();
+
+  const { value: originalDebt, updateCount } = await E(
+    debtNotifier,
+  ).getUpdateSince();
+  t.deepEqual(originalDebt, maxLoan);
+
+  // @ts-ignore
+  periodUpdater.updateState(5);
+
+  const { value: debtCompounded1, updateCount: updateCount1 } = await E(
+    debtNotifier,
+  ).getUpdateSince(updateCount);
+  t.deepEqual(debtCompounded1, loanKit.amountMath.make(40020));
+
+  // skip ahead a notification
+  // @ts-ignore
+  periodUpdater.updateState(15);
+
+  // both debt notifications are received
+  const { value: debtCompounded2, updateCount: updateCount2 } = await E(
+    debtNotifier,
+  ).getUpdateSince(updateCount1);
+  t.is(15, await E(borrowFacet).getLastCalculationTimestamp());
+  t.deepEqual(debtCompounded2, loanKit.amountMath.make(40040));
+  const { value: debtCompounded3 } = await E(debtNotifier).getUpdateSince(
+    updateCount2,
+  );
+  t.deepEqual(debtCompounded3, loanKit.amountMath.make(40060));
+});
+
```

Comment:
IIUC, I think @katelynsills race condition worry is serious, even if it doesn't happen in practice. It may be fundamental to the problem, but I don't know one way or the other. 

IIUC correctly again (I'm not sure I do), I agree with @Chris-Hibbert that the lossiness of the notifier does not create the problem. Rather, it is the unpredicability of the delay. AFAICT, the exact same race condition would have occurred with the asyncIterable subscription. If I'm misunderstanding and that one would not have this race condition, then I suspect the notifier does not either. But I'd like to understand this before concluding anything.

Reply:
I think you're right about the fundamentals. Hmmm. Thinking out loud:

Our computational model has no realtime guarantees. But finance has, and must have, hard realtime requirements --- deadlines. This mismatch is fundamental. Bizarrely, the long inter-blocktime delays of blockchain may actually be clarifying here. Compared to the inter-blocktime delay, each allowed turn is almost instantaneous. And turn that is not exhausts the meter and did not happen. Thus, eventually we should be able to reason about number-of-blocks in much the same way that E programmers reason about number-of-round-trips (we we also need to reason about). Predictable inter-block delays is like a synchronizing clock signal. Given a bunch of assumptions about what component runs of which chain, and what the round trip and block delay counts are on each of those chains, and how many of each delay is needed to decide to make a payment and to make the payment, one may come to have confidence that a given arrangement for making a payment will meet the deadline set by the financial arrangement.

Way too many ifs and too much mechanism for what we're doing now or any time soon. But it is nice to know there's a way to approach this without needing to pretend that JS itself has any realtime properties. Turn counts, crank counts, block counts, and round trip counts should all reasonably straightforward to analyze for many concrete arrangements.

Attn @warner @dtribble

Reply:
A lot to respond to here. Probably doesn't belong so deep in a thread. I'll have separate replies on these things

re "lossy notifier" is a misnomer and we should change. Periodic timers are lossy in that you might miss one (e.g., because you took too long or the system was too busy). Notifiers are only lossy in that they miss indivdual updates, but they don't lose actual data. A different term would be good.

Reply:
Loan fees should be in terms of some notion of time. Blockheight is convenient for that, but is not clock time. If we add time consensus to blocks, we could not use it for the current block, but could use it for the next block. So the timer service schedules tasks in the next block that met their deadline in this block.

Reply:
My assumption about a payoff offer is that it would use offer safety for: "pay at most my debt for at least my current collateral".  If the debt had increased in the meantime, then the contract would not release collateral so the client would not lose the money. The payoff would be rejected.

If fees are time-based, then they should be brought current before any state change. Thus the notifier brings it current and then tests for liquidation.  A balance query brings it current and returns the balance. Paying it off brings it current and then applies the payment. Offer safety should be such that, if the payoff isn't enough, the loan cannot collect it.

If fees are event-based (or epoch-based) such that you owe them only if a fee coupon is presented, then all the race is external to the contract. If the payoff arrives before a fee request is presented, then the account is closed and the fee "offer" will get rejected. There no race issue within the contract at all.

Reply:
"hard real time" means millisecond response times and precision. Important for device control.

Finance is "very soft real time" generally. Traditional finance is generally only accurate to the day. That's pretty soft :).  Some such system seem more precise because they go fast, but they only care about order. I would very surprised if there are trading systems that let you schedule transaction at finer granularity than 15 minutes.

Reply:
> "lossy notifier" is a misnomer and we should change. 

Agreed. I think there are no vestiges of this remaining in this PR.

> Loan fees should be in terms of some notion of time. 

The fees are in terms of `interestRate` and `interestPeriod`.  The contract doesn't poll the clock, it uses a timer notifier, which means it gets notified after (most) repetitions of `interestPeriod`. We could check the clock when the borrower tries to repay, but I think that's overkill. If we expected the Notifier to actually be lossy, this might be called for, but there shouldn't be missed updates at the periodicity we're intending here.

> My assumption about a payoff offer is that it would use offer safety for: "pay at most my debt for at least my current collateral". If the debt had increased in the meantime, then the contract would not release collateral so the client would not lose the money. The payoff would be rejected.

That seems right.

> Finance is "very soft real time" generally. 

This discussion seems right.